fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -2089,6 +2093,210 @@ private boolean isStateCacheableInternal() {
+     * </p>
+     * <ul>
+     *     <li>
+     *         Because all operations to a single record store must read the store header, every time the
+     *         value is updated, any concurrent operations to the store will fail with a
+     *         {@link com.apple.foundationdb.record.provider.foundationdb.FDBExceptions.FDBStoreTransactionConflictException FDBStoreTransactionConflictException}.
+     *         Therefore, this should only ever be used for data that are mutated at a very low rate.
+     *     </li>
+     *     <li>
+     *         As this data must be read every time the record store is created (which is at least once per
+     *         transaction), the value should not be too large. There is not at the moment a hard limit applied, but
+     *         a good rule of thumb may be to keep the total size from all user fields under a kilobyte.
+     *     </li>
+     * </ul>
+     *
+     * <p>
+     * The value of these fields are simple byte arrays, so the user is free to choose their own serialization
+     * format for the value included in this field. One recommended strategy is for the user to use Protobuf to
+     * serialize a message with possibly multiple keys and values, each with meaning to the user, and then to
+     * write it to a single user field (or a small number if appropriate). This decreases the total size of
+     * the user fields when compared to something like storing one user-field for each field (as the Protobuf
+     * serialization format is more compact), and it allows the user to follow standard Protobuf evolution
+     * guidelines as these fields evolve.
+     * </p>
+     *
+     * <p>
+     * Once set, the value of these fields can be retrieved by calling {@link #getHeaderUserField(String)}. They
+     * can be cleared by calling {@link #clearHeaderUserFieldAsync(String)}.
+     * </p>
+     *
+     * <p>
+     * Using this feature also requires that the record store be on format version {@link #HEADER_USER_FIELDS_FORMAT_VERSION}
+     * or higher. There is no change to the on-disk format version from the previous format version except that
+     * extra fields in the store header were added. No data migration is necessary to upgrade to that version if
+     * an existing store is on the previous format version, but the new version is used to prevent the user from serializing
+     * data and then not being able to read it from instances running an older version of the Record Layer.
+     * </p>
+     *
+     * @param userField the name of the user-settable field to write
+     * @param value the value to set the field to
+     * @return a future that will be ready when setting the field has completed
+     * @see #getHeaderUserField(String)
+     */
+    @Nonnull
+    public CompletableFuture<Void> setHeaderUserFieldAsync(@Nonnull String userField, @Nonnull ByteString value) {
+        return updateStoreHeaderAsync(storeHeaderBuilder -> {
```

Comment:
The way this is written, if you set a user field it won't be visible from `getHeaderUserField()` until the transaction has been committed.  Perhaps this should have a javadoc to this effect, or this method should actually update the in-memory version prior to returning the future?   I think the former probably makes more sense.